server/game/cards/06-CotE/MiyaLibrary.js

```diff
@@ -0,0 +1,56 @@
+const DrawCard = require('../../drawcard.js');
+const { CardTypes } = require('../../Constants');
+// const AbilitDsl = require('../../abilitydsl');
+
+class MiyaLibrary extends DrawCard {
+    setupCardAbilities() {
+        this.action({
+            title: 'Replace Miya Library for a faceup imperial character',
+            condition: context => context.player.dynastyDeck.size() > 0,
+            effect: 'Search the top four card for your dynasty deck for an imperial character',
+            handler: context => {
+                this.game.promptWithHandlerMenu(context.player, {
+                    activePromptTitle: 'select an imperial character to replace miya library',
+                    context: context,
+                    cardCondition: card => card.hasTrait('imperial') && card.getType() === CardTypes.Character,
+                    cards: context.player.dynastyDeck.first(4),
+                    choices: ['Do not replace Miya Library'],
+                    handlers: [() => this.miyaLibraryPrompt(context, context.player.dynastyDeck.first(4), [], 'Select the card you would like to place on top of your dynasty deck')],
+                    cardHandler: (card) => {
+                        let choices = context.player.dynastyDeck.first(4);
+                        if(card.hasTrait('imperial') && card.getType() === CardTypes.Character) {
+                            context.player.moveCard(card, context.source.location);
+                            card.facedown = false;
+                            choices.splice(choices.indexOf(card), 1);
+                            choices.push(context.source);
+                        }
+                        this.miyaLibraryPrompt(context, choices, [], 'Select the card you would like to place on top of your dynasty deck');
+                    }
```

Comment:
I think this should be:
`cardHandler: (card) => {
                         if(card.hasTrait('imperial') && card.getType() === CardTypes.Character) {
                            context.player.moveCard(card, context.source.location);
                            card.facedown = false;
                            context.player.moveCard(context.source, Locations.DynastyDeck);
                        }
                        this.miyaLibraryPrompt(context, context.player.dynastyDeck.first(4), [], 'Select the card you would like to place on top of your dynasty deck');
                    }`